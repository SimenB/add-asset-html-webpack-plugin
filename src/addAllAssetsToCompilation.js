import path from 'path';
import crypto from 'crypto';
import pEachSeries from 'p-each-series';
import micromatch from 'micromatch';
import handleUrl from './handleUrl';

function ensureTrailingSlash(string) {
  if (string.length && string.substr(-1, 1) !== '/') {
    return `${string}/`;
  }

  return string;
}

// Copied from html-webpack-plugin
function resolvePublicPath(compilation, filename) {
  /* istanbul ignore else */
  const publicPath =
    typeof compilation.options.output.publicPath !== 'undefined'
      ? compilation.options.output.publicPath
      : path.relative(path.dirname(filename), '.'); // TODO: How to test this? I haven't written this logic, unsure what it does

  return ensureTrailingSlash(publicPath);
}

function resolveOutput(compilation, addedFilename, outputPath) {
  if (outputPath && outputPath.length) {
    /* eslint-disable no-param-reassign */
    compilation.assets[`${outputPath}/${addedFilename}`] =
      compilation.assets[addedFilename];
    delete compilation.assets[addedFilename];
    /* eslint-enable */
  }
}

async function addFileToAssets(
  compilation,
  htmlPluginData,
  {
    filepath,
    typeOfAsset = 'js',
    includeSourcemap = true,
    hash = false,
    publicPath,
    outputPath,
    files = [],
  },
) {
  if (!filepath) {
    const error = new Error('No filepath defined');
    compilation.errors.push(error);
    return Promise.reject(error);
  }

  const fileFilters = Array.isArray(files) ? files : [files];

  if (fileFilters.length > 0) {
    const shouldSkip = !fileFilters.some(file =>
      micromatch.isMatch(htmlPluginData.outputName, file),
    );

    if (shouldSkip) {
      return Promise.resolve(null);
    }
  }

  const addedFilename = await htmlPluginData.plugin.addFileToAssets(
    filepath,
    compilation,
  );

  let suffix = '';
  if (hash) {
    const md5 = crypto.createHash('md5');
    md5.update(compilation.assets[addedFilename].source());
    suffix = `?${md5.digest('hex').substr(0, 20)}`;
  }

  const resolvedPublicPath =
    typeof publicPath === 'undefined'
      ? resolvePublicPath(compilation, addedFilename)
      : ensureTrailingSlash(publicPath);
  const resolvedPath = `${resolvedPublicPath}${addedFilename}${suffix}`;

  htmlPluginData.assets[typeOfAsset].unshift(resolvedPath);

  resolveOutput(compilation, addedFilename, outputPath);

  if (includeSourcemap) {
    const addedMapFilename = await htmlPluginData.plugin.addFileToAssets(
      `${filepath}.map`,
      compilation,
    );
    resolveOutput(compilation, addedMapFilename, outputPath);
  }

  return Promise.resolve(null);
}

// Visible for testing
export default async function(assets, compilation, htmlPluginData) {
  const handledAssets = await handleUrl(assets);
  await pEachSeries(handledAssets, asset =>
    addFileToAssets(compilation, htmlPluginData, asset),
  );
  return htmlPluginData;
}
