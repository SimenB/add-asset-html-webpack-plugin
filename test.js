/* eslint import/no-extraneous-dependencies: ["error", { "devDependencies": true }] */
/* eslint-env jest */

import path from 'path';
import assert from 'power-assert';
import sinon from 'sinon';
import Promise from 'bluebird';
import AddAssetHtmlPlugin from './src/index';
import addAllAssetsToCompilation from './src/addAllAssetsToCompilation';

const pluginMock = {
  plugin: {
    addFileToAssets: filename => Promise.resolve(path.basename(filename)),
  },
};

test('assets should always be an array', () => {
  assert(Array.isArray(new AddAssetHtmlPlugin({}).assets));
  assert(Array.isArray(new AddAssetHtmlPlugin([]).assets));
  assert(Array.isArray(new AddAssetHtmlPlugin().assets));
});

test('assets should should be reversed', () => {
  assert.deepEqual(new AddAssetHtmlPlugin(['a', 'b']).assets, ['b', 'a']);
});

test.concurrent('should invoke callback on success', async () => {
  const callback = sinon.stub();

  await addAllAssetsToCompilation([], {}, pluginMock, callback);

  assert(callback.calledOnce);
  assert(callback.calledWithExactly(null, pluginMock));
});

test.concurrent('should invoke callback on error', async () => {
  const callback = sinon.stub();
  const compilation = { errors: [] };

  await addAllAssetsToCompilation([{}], compilation, pluginMock, callback);

  assert(compilation.errors.length === 1);
  assert(compilation.errors[0].message === 'No filepath defined');

  assert(callback.calledOnce);
  assert(callback.calledWithExactly(compilation.errors[0], pluginMock));
});

test.concurrent("should add file using compilation's publicPath", async () => {
  const callback = sinon.stub();
  const compilation = { options: { output: { publicPath: 'vendor/' } } };
  const pluginData = Object.assign({ assets: { js: [], css: [] } }, pluginMock);

  await addAllAssetsToCompilation([{ filepath: path.join(__dirname, 'my-file.js') }], compilation, pluginData, callback);

  assert.deepEqual(pluginData.assets.css, []);
  assert.deepEqual(pluginData.assets.js, ['vendor/my-file.js']);

  assert(callback.calledOnce);
  assert(callback.calledWithExactly(null, pluginData));
});

test.concurrent('should used passed in publicPath', async () => {
  const callback = sinon.stub();
  const compilation = { options: { output: { publicPath: 'vendor/' } } };
  const pluginData = Object.assign({ assets: { js: [], css: [] } }, pluginMock);

  await addAllAssetsToCompilation([{ filepath: 'my-file.js', publicPath: 'pp' }], compilation, pluginData, callback);

  assert.deepEqual(pluginData.assets.css, []);
  assert.deepEqual(pluginData.assets.js, ['pp/my-file.js']);

  assert(callback.calledOnce);
  assert(callback.calledWithExactly(null, pluginData));
});

// No idea what this does, actually... Coverage currently hits it, but the logic is untested.
test('should handle missing `publicPath`');

test.concurrent('should add file missing "/" to public path', async () => {
  const callback = sinon.stub();
  const compilation = { options: { output: { publicPath: 'vendor' } } };
  const pluginData = Object.assign({ assets: { js: [], css: [] } }, pluginMock);

  await addAllAssetsToCompilation([{ filepath: 'my-file.js' }], compilation, pluginData, callback);

  assert.deepEqual(pluginData.assets.css, []);
  assert.deepEqual(pluginData.assets.js, ['vendor/my-file.js']);

  assert(callback.calledOnce);
  assert(callback.calledWithExactly(null, pluginData));
});

test.concurrent('should add sourcemap to compilation', async () => {
  const callback = sinon.stub();
  const addFileToAssetsStub = sinon.stub();
  const compilation = { options: { output: {} } };
  const pluginData = { assets: { js: [], css: [] }, plugin: { addFileToAssets: addFileToAssetsStub } };
  addFileToAssetsStub.returns(Promise.resolve('my-file.js'));

  await addAllAssetsToCompilation([{ filepath: 'my-file.js' }], compilation, pluginData, callback);

  assert.deepEqual(pluginData.assets.css, []);
  assert.deepEqual(pluginData.assets.js, ['my-file.js']);

  assert(callback.calledOnce);
  assert(callback.calledWithExactly(null, pluginData));

  assert(addFileToAssetsStub.calledTwice);
  assert(addFileToAssetsStub.getCall(0).args[0] === 'my-file.js');
  assert(addFileToAssetsStub.getCall(1).args[0] === 'my-file.js.map');
});

test.concurrent('should skip adding sourcemap to compilation if set to false', async () => {
  const callback = sinon.stub();
  const addFileToAssetsStub = sinon.stub();
  const compilation = { options: { output: {} } };
  const pluginData = { assets: { js: [], css: [] }, plugin: { addFileToAssets: addFileToAssetsStub } };
  addFileToAssetsStub.returns(Promise.resolve('my-file.js'));

  await addAllAssetsToCompilation([{ filepath: 'my-file.js', includeSourcemap: false }], compilation, pluginData, callback);

  assert.deepEqual(pluginData.assets.css, []);
  assert.deepEqual(pluginData.assets.js, ['my-file.js']);

  assert(callback.calledOnce);
  assert(callback.calledWithExactly(null, pluginData));

  assert(addFileToAssetsStub.calledOnce);
  assert(addFileToAssetsStub.getCall(0).args[0] === 'my-file.js');
});

test.concurrent('should include hash of file content if option is set', async () => {
  const callback = sinon.stub();
  const compilation = {
    options: { output: {} },
    assets: { 'my-file.js': { source: () => 'some source code is cool to have;' } },
  };
  const pluginData = Object.assign({ assets: { js: [], css: [] } }, pluginMock);

  await addAllAssetsToCompilation([{ filepath: 'my-file.js', hash: true }], compilation, pluginData, callback);

  assert.deepEqual(pluginData.assets.css, []);
  assert.deepEqual(pluginData.assets.js, ['my-file.js?5329c141291f07ab06c6']);

  assert(callback.calledOnce);
  assert(callback.calledWithExactly(null, pluginData));
});

test.concurrent('should add to css if `typeOfAsset` is css', async () => {
  const callback = sinon.stub();
  const compilation = {
    options: { output: {} },
    assets: { 'my-file.js': { source: () => 'some source code is cool to have;' } },
  };
  const pluginData = Object.assign({ assets: { js: [], css: [] } }, pluginMock);

  await addAllAssetsToCompilation([{ filepath: 'my-file.css', typeOfAsset: 'css' }], compilation, pluginData, callback);

  assert.deepEqual(pluginData.assets.css, ['my-file.css']);
  assert.deepEqual(pluginData.assets.js, []);

  assert(callback.calledOnce);
  assert(callback.calledWithExactly(null, pluginData));
});

test.concurrent('should replace compilation assets key if `outputPath` is set', async () => {
  const callback = sinon.stub();
  const source = { source: () => 'test' };
  const addFileToAssetsMock = (filename, compilation) => {
    const name = path.basename(filename);
    compilation.assets[name] = source; // eslint-disable-line no-param-reassign
    return Promise.resolve(name);
  };
  const compilation = {
    options: { output: {} },
    assets: {},
  };
  const pluginData = { assets: { js: [], css: [] }, plugin: { addFileToAssets: addFileToAssetsMock } };

  await addAllAssetsToCompilation([{ filepath: 'my-file.js', outputPath: 'assets' }], compilation, pluginData, callback);

  assert.deepEqual(pluginData.assets.css, []);
  assert.deepEqual(pluginData.assets.js, ['my-file.js']);

  assert(compilation.assets['my-file.js'] === undefined);
  assert.deepEqual(compilation.assets['assets/my-file.js'], source);
  assert(compilation.assets['my-file.js.map'] === undefined);
  assert.deepEqual(compilation.assets['assets/my-file.js.map'], source);
});
