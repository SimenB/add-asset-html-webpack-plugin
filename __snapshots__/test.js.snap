// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter option should exclude some files 1`] = `
Object {
  "css": Array [],
  "js": Array [],
}
`;

exports[`filter option should include some files 1`] = `
Object {
  "css": Array [],
  "js": Array [
    "vendor/my-file.js",
  ],
}
`;

exports[`filter option should include some files with string option 1`] = `
Object {
  "css": Array [],
  "js": Array [
    "vendor/my-file.js",
  ],
}
`;

exports[`should add file missing "/" to public path 1`] = `
Object {
  "css": Array [],
  "js": Array [
    "vendor/my-file.js",
  ],
}
`;

exports[`should add file using compilation's publicPath 1`] = `
Object {
  "css": Array [],
  "js": Array [
    "vendor/my-file.js",
  ],
}
`;

exports[`should add sourcemap and gzipped files to compilation 1`] = `
Object {
  "css": Array [],
  "js": Array [
    "my-file.js",
  ],
}
`;

exports[`should add to css if \`typeOfAsset\` is css 1`] = `
Object {
  "css": Array [
    "my-file.css",
  ],
  "js": Array [],
}
`;

exports[`should include hash of file content if option is set 1`] = `
Object {
  "css": Array [],
  "js": Array [
    "my-file.js?5329c141291f07ab06c6",
  ],
}
`;

exports[`should invoke callback on error 1`] = `
Array [
  [Error: No filepath defined],
]
`;

exports[`should reject on error 1`] = `
Array [
  [Error: No filepath defined],
]
`;

exports[`should replace compilation assets key if \`outputPath\` is set 1`] = `
Object {
  "css": Array [],
  "js": Array [
    "some-file.js",
  ],
}
`;

exports[`should skip adding sourcemap and gzipped files to compilation if set to false 1`] = `
Object {
  "css": Array [],
  "js": Array [
    "my-file.js",
  ],
}
`;

exports[`should use the hash from HtmlWebpackPlugin if option is set 1`] = `
Object {
  "css": Array [],
  "js": Array [
    "my-file.js?testHash",
  ],
}
`;

exports[`should used passed in publicPath 1`] = `
Object {
  "css": Array [],
  "js": Array [
    "pp/my-file.js",
  ],
}
`;
